content/posts/codegen.md

```diff
@@ -0,0 +1,143 @@
+
+Writing support for each of these resource types looks pretty
+similar. They all follow the same rough logic:
+
+- A user needs a cloud resource to exist in a certain manner
+- The tool calls that cloud resource's GET API to see if the resource exists.
+- If it doesn't, the tool calls the resource type's Create API.
+- If the resource does exist, the tool checks to see if the user's preferences
+  differ from whatever the resource currently has.
+- If it does, the update API gets called.
+
+You might first think that we need a function to handle this. The function could
+take in the user's intention and links to Get/Create/Update APIs. But, as time
+goes on, you realize that some resources handle the creation process slightly
+different from others. Then, every resource type expresses their current state
+(from the GET API) slightly differently from each other. That makes handling the
+"check if user's preferences are different" phase slightly different for each
+resource. As you go on, you realize that parts of each resource type's code will
+be identical to all the others and other parts will be radically different.
+
+Each resource type also needs to have a file with a handful of boilerplate
+methods defined. No amount of functional or class abstraction can change the
+fact that our team needs to write out all of those boilerplate methods...and
+that's before the fact that we realize that the boilerplate isn't just
+copy-paste.
+
+Doing this once or twice is interesting! Doing it a dozen or so times feels
+monotonous, especially once you realize that you made the same bug in all dozen
+resource types and have to make the same fix twelve times. Now, imagine doing
+it a hundred times.
+
+Code generation lets us create a source-of-truth for all this boilerplate.
+
+## How to Generate Code
+
+There's several different approaches to code generation.
+
+- **Compiler Magic:** You could programmatically build out the proper Abstract
+  Syntax Tree and then rely on the compiler to either output the AST as code or
+  just run the AST directly. I suspect that most people handle code generation
+  with some amount of this. That's not actually the method I'm going to be
+  talking about. While you don't have to have a compiler background to do this,
+  it does require knowledge of your language's internals. Testing will be more
+  straightforward of any of these.
+
+- **AI/ML:** Oh, please no! Take that witchcraft away and back to the godless
+  land from which it was brewed. Seriously though, it's really hard to have any
+  insight into what a ML model is doing or how to debug it. Your training data
+  is probably going to look a lot like your final product, which sort of
+  defeats the purpose of automation.
+
+While many people will take one of these approaches (especially the first one),
+we've chosen not to. We actually take a third approach: templating.
+
+## Templated Approach
+
+The approach we take is essentially Mad Libs.
```

Comment:
idea: I think it may be useful to start this section off by talking about templating being used with strings in languages, like C# string interpolation or C++ `std::format`, as these are things many people may have experience with. you could then motivate that you're applying the same principals but at the scale of entire files and more.